### PR TITLE
In tool selection

### PR DESCRIPTION
**Pull request by Eneroth3**
Adding example showing in-tool selection.

This example has gotten quite big though and shows other things too. Maybe it should be relabeled as a Move tool example, and the Readme be elaborated a bit more?

There is also a hack to get around a to me new SU bug related to InputPoint cloning and picking. I haven't looked into this bug further, as I'm getting quite far away from the original purpose of this branch, providing a code example for the UX guidelines.

### CODE REVIEW

**Comment by Eneroth3** [comment]
- [ ] Squash

File: examples/05_tool_inference_lock/ex_tool_inference_lock/main.rb
Comment (thomthom) [inline examples/05_tool_inference_lock/ex_tool_inference_lock/main.rb:133]: What's this change for?

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [inline examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: A more accurate name could be `@preselected`?

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [inline examples/99_in_tool_selection/ex_in_tool_selection/main.rb:48]: What does this variable do/represent?

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [inline examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: I would have expected these to be in `initialize` (apart from the call to `update_ui`).

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [inline examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: This comment is confusing me. We aren't picking at all at this if branch. Can you clarify?

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [inline examples/99_in_tool_selection/ex_in_tool_selection/main.rb:126]: No need to reset anything if there wasn't a drag?

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [inline examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: I would prefer to see `.vector_to` here instead of the minus operator as it much more explicit. Various 3d libraries have various meaning for the math operators on 3d geometry objects - which can be a source of confusion.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [inline examples/99_in_tool_selection/ex_in_tool_selection/main.rb:168]: I would like to see a comment here on what's going on. Seems like you are simulating mouse input?

It seems strange to me to generate an input point from text input though... This is one type of user input driving another form of user input to drive the state of the tool.

```
keyboard -> mouse -> tool-state
```

I'd prefer to see something like:

```
keyboard -+
          +-->tool-state
   mouse -+
```

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [inline examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: "linked"?

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [inline examples/99_in_tool_selection/ex_in_tool_selection/main.rb:256]: Worth noting that observers will not still while the operation is open. In case someone tried to use observers along with their tool while doing a long operation like this.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [inline examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: "frames"?

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [inline examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: Also here I'd like to see `.vector_to`

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [inline examples/99_in_tool_selection/ex_in_tool_selection/main.rb:298]: "quite strange behavior" could do with some explanation here. People read this to learn and will puzzle to what this means. (I do too.)

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [inline examples/99_in_tool_selection/ex_in_tool_selection/main.rb:321]: Is this described in the UX guidelines? This sounds interesting, though I'd like to learn more what behaviour this relate to.

**Review by thomthom — CHANGES_REQUESTED** [review]
I think this example looks good. Walks through a lot of details that makes for the SketchUp UX.

Had some comments on areas I felt needed more info/explaination.

File: examples/05_tool_inference_lock/ex_tool_inference_lock/main.rb
Comment (Eneroth3) [reply examples/05_tool_inference_lock/ex_tool_inference_lock/main.rb:133]: Mistake from previous PR. This rescue was supposed to be moved from the method as a whole to the Length parsing specifically, but I forgot to remove this code.

This should probably not be squashed with the rest as it is an unrelated change.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: Switching to `@preselected`

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb:48]: Added comment

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb:126]: You could set `@mouse_down` to nil or ORIGIN, but I don't know if it's needed.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb:168]: You could have a separate instance variable for a length and point and have logic to determine what value to use when, but I'm thinking it's simpler to just have a single value.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb:256]: I don't know if we should introduce that topic here. The reader may not know what observers are and get more confused by it. Also i think it's a quite specific case to use observer along with your custom tools.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: I don't really know what to call it. Mouse move perhaps? But "a move" could be interpreted as being from you start moving the mouse until you stop it, not the numerous individual events.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: I'm more used to this syntax but I don't know if one is better than the other. My brain likes when there are fewer tokens to parse.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: I don't know how to describe this bug. With the original code the reference input point countinosuly copy the value from the point that is being picked, causing the preview line to disappear and the VCB to show values close to 0. There is however a small delay, maybe from the previous pick, and the two InputPoints don't get the exact same position.

I don't know if it's a memory issue where both InputPoints get mixed up or what's causing it. This should be investigated, but probably not blocking the release of the UX guidelines.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb:298]: Clarified the comment.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb:321]: Clarified the comment. This has to do with key event triggering a UI update, as they could change what position the InputPoints should be picking based on inference locking.

This might be outside the scope of the UX guidlines. It's more about preventing a clearly bugged behavior than a UX design choice.

For the record SketchUp some similar bug where pressing Shift to write certain character on certain keyboards (`<` and `>` I believe) flushes the text the user is writing, making it impossible to enter certain characters.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb:126]: Ok, just checking as I often find tool state needs a few members to be reset.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb:256]: Fair point.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: I think `@last_mouse_position` is sufficient. "frames" threw me off a bit here.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb:321]: Seems to me the behaviour affects user experience. That if an extension doesn't do this the VCB gets out of sync?

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: `.vector_to` makes it obvious that you get a vector in return. `-` is not so clear in it's intent. Being unfamiliar with the API or being used to a different 3D API this could be read as subtracting the members and returning a `Point3d`.

`.vector_to` is unambiguous to what it does which is a benefit when used with examples/tutorials. If this had not been an example or tutorial I wouldn't objected to either variant.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb:168]: Could you add some comment here to explain what is happening. I'm still not exactly clear on what we're doing where. Why we need to call `do_move` and `end_move` in immediate succession.

Could `end_move` that the mouse event calls call another function that performs the move, and this onUserText could then use directly?

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: Yea, if you could capture the steps to reproduce, along with a couple of screenshots to illustrate what happens into a bug report in our issue tracker that would be good. I'm curious to see what is doing on here.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb:321]: There is already a  guideline saying you should use the VCB when applicable. I think that one already implies the ability to write to the VCB.

Comment (thomthom) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb:321]: I was referring to what you described as buggy behaviour. If extensions aren't doing what you do here, the VCB values might not represent the state of the tool, right?

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: Changing to vector_to

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb:168]: When you enter a value you also finish the operation, not just updating the current position. Adding comments to clarify.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb:321]: It would still represent the state of the tool but you wouldn't be able to type anything as the key event would set the VCB value.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (thomthom) [inline examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: I didn't understand this: "This affects the mouse cursor."

**Comment by thomthom** [comment]
I think we're almost complete, only a couple of comments that would be nice to clarify. I "Resolved" the comments that is no longer an issue and left open the two that would be nice to tweak.

Is that a workflow that works ok here on GitHub? Use comments and "Resolve Conversation" similar to how we use Tasks on JIRA? This does mean we need to be more explicit to comments that expect changes and those that doesn't. Maybe insert a checkbox to comments where we want to see changes? (if that can be done in code-comments...)

File: examples/99_in_tool_selection/README.md
Comment (thomthom) [inline examples/99_in_tool_selection/README.md:6]: - [ ] Testing a code comment with a checkbox. Is this something that can work in clarifying what should be resolved before merge? Like JIRA?

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: Changing to "This affects what mouse cursor the tool uses."

File: examples/99_in_tool_selection/README.md
Comment (Eneroth3) [reply examples/99_in_tool_selection/README.md:6]: I don't think checkboxes gives any warnings before closing, but maybe can be used manually before resolving conversations.

File: examples/99_in_tool_selection/ex_in_tool_selection/main.rb
Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: I think this might be a quite deep rabbit hole but I make a simple issue that references the code example, so this work doesn't block publishing the UX Guidelines.

Comment (Eneroth3) [reply examples/99_in_tool_selection/ex_in_tool_selection/main.rb]: Found it! (https://github.com/SketchUp/api-issue-tracker/issues/618)

Rewritten the comment to better communicate what is done and why.

File: examples/99_in_tool_selection/README.md
Comment (thomthom) [reply examples/99_in_tool_selection/README.md:6]: It's possible to set up checks for GitHub PRs. So we might be able to setup a script that checks for checkboxes. Might even be an existing solution we can use. Worth a look at some point as I expect we'll be using github more going forward.

**Review by thomthom — APPROVED** [review]
Nice!

**Pull request merged**
